Encode enums as "raw" ints, rather than a node

SquidDev opened the pull request:
This avoids one layer of indirection in serialisation/deserialisation, slightly reducing the allocation rate and CPU time.

## Performance
Numbers were taken using GHC 9.4.3 and testing against an already type-checked version of https://github.com/plt-amy/1lab/commit/bb69dd0df729135e978f302984c0d65abbcfb2b2. Agda was run under `perf stat` and with `--profile=internal +RTS -s -RTS`

<details><summary>master (<code>6206624eef597afea711120110c219e266562a26</code>)</summary>

```
Total                      13,564ms
Miscellaneous                   4ms
Deserialization             9,422ms (13,408ms)
Deserialization.Compaction  3,986ms
Import                        116ms
Parsing                        35ms
  12,231,231,312 bytes allocated in the heap
   4,972,604,800 bytes copied during GC
     619,710,368 bytes maximum residency (24 sample(s))
       1,194,080 bytes maximum slop
            1055 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      2551 colls,     0 par    5.415s   5.426s     0.0021s    0.0207s
  Gen  1        24 colls,     0 par    0.336s   0.337s     0.0140s    0.0350s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time    7.850s  (  7.797s elapsed)
  GC      time    5.751s  (  5.763s elapsed)
  EXIT    time    0.001s  (  0.010s elapsed)
  Total   time   13.604s  ( 13.571s elapsed)

  Alloc rate    1,558,042,514 bytes per MUT second

  Productivity  57.7% of total user, 57.5% of total elapsed


 Performance counter stats for '/home/squid/Documents/software/local-copies/agda/dist-newstyle/build/x86_64-linux/ghc-9.4.3/Agda-2.6.4/build/agda/agda _build/all-pages.agda --profile=internal +RTS -s -RTS':

         13,625.22 msec task-clock:u                     #    1.001 CPUs utilized
                 0      context-switches:u               #    0.000 /sec
                 0      cpu-migrations:u                 #    0.000 /sec
           411,508      page-faults:u                    #   30.202 K/sec
    52,848,740,179      cycles:u                         #    3.879 GHz                         (83.13%)
     2,287,938,016      stalled-cycles-frontend:u        #    4.33% frontend cycles idle        (83.32%)
    21,536,235,871      stalled-cycles-backend:u         #   40.75% backend cycles idle         (83.40%)
    69,248,026,539      instructions:u                   #    1.31  insn per cycle
                                                  #    0.31  stalled cycles per insn     (83.38%)
    13,113,978,085      branches:u                       #  962.478 M/sec                       (83.36%)
       295,320,153      branch-misses:u                  #    2.25% of all branches             (83.40%)

      13.616020670 seconds time elapsed

      12.790328000 seconds user
       0.858570000 seconds sys
```

</details>

<details><summary>This commit</summary>

```
Total                      13,388ms           
Miscellaneous                  11ms           
Deserialization             9,041ms (13,251ms)
Deserialization.Compaction  4,209ms           
Import                        108ms           
Parsing                        17ms           
  12,103,302,296 bytes allocated in the heap
   4,930,846,360 bytes copied during GC
     612,042,352 bytes maximum residency (22 sample(s))
       1,542,416 bytes maximum slop
            1087 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      2523 colls,     0 par    5.377s   5.402s     0.0021s    0.0169s
  Gen  1        22 colls,     0 par    0.371s   0.372s     0.0169s    0.0716s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time    7.681s  (  7.618s elapsed)
  GC      time    5.747s  (  5.774s elapsed)
  EXIT    time    0.001s  (  0.007s elapsed)
  Total   time   13.431s  ( 13.400s elapsed)

  Alloc rate    1,575,659,199 bytes per MUT second

  Productivity  57.2% of total user, 56.8% of total elapsed


 Performance counter stats for '/home/squid/Documents/software/local-copies/agda/dist-newstyle/build/x86_64-linux/ghc-9.4.3/Agda-2.6.4/build/agda/agda _build/all-pages.agda --profile=internal +RTS -s -RTS':

         13,473.99 msec task-clock:u                     #    1.002 CPUs utilized             
                 0      context-switches:u               #    0.000 /sec                      
                 0      cpu-migrations:u                 #    0.000 /sec                      
           430,910      page-faults:u                    #   31.981 K/sec                     
    51,990,949,231      cycles:u                         #    3.859 GHz                         (83.41%)
     2,059,375,865      stalled-cycles-frontend:u        #    3.96% frontend cycles idle        (83.41%)
    21,527,709,644      stalled-cycles-backend:u         #   41.41% backend cycles idle         (83.48%)
    68,180,447,464      instructions:u                   #    1.31  insn per cycle            
                                                  #    0.32  stalled cycles per insn     (83.44%)
    12,927,895,239      branches:u                       #  959.470 M/sec                       (83.40%)
       294,750,935      branch-misses:u                  #    2.28% of all branches             (82.85%)

      13.452032460 seconds time elapsed

      12.695894000 seconds user
       0.786864000 seconds sys
```

</details>

There's about a 1% reduction in allocations and instruction count. Running with `+RTS -A128 -RTS` shows similar improvements, though obviously with less GC pressure.

My computer is really not set up for benchmarking, and so the CPU time varies wildly. The above numbers are a best of 5 runs.

## Misc notes
`EmbPrj` does include a default implementation for `Enum`s already, which is used in a couple of places. There was some discussion about using them more widely in #5977 - I know there were concerns about forgetting to bump the interface version when changing enums.

It does feel a shame not to use the default methods here, but also I don't really have any good ideas about enforcing bumping the interface version - mostly noting this for context!

andreasabel commented on src/full/Agda/TypeChecking/Serialise/Instances.hs:
```diff
@@ -28,8 +28,5 @@ instance EmbPrj Interface where
-      where interface a b c = Interface a b c . toImportedModules
-    valu _ = malformed
+  value = valueN interface
+    where interface a b c = Interface a b c . toImportedModules
```
Is this change related to the topic of the PR or is it just some simplification you noticed while working on the code?  I fail to see the connection...

andreasabel replied:
There are more occurrences of such refactorings.
You seem to have simplified the implementation of `value` for single-constructor data types by skipping `vcase`.

SquidDev replied:
`valueN` still uses `vcase` under the hood (so should still have the same memoization/sharing behaviour), I just find it forms a more natural dual to `icodeN'`. Do appreciate it is a separate change to the enum one though, so happy to revert!

andreasabel replied:
I welcome this change.  
Could you put it in a separate commit with a descriptive commit message?  Then we would merge this with two separate commits.  This is my suggestion.

andreasabel reviewed (approved):
Thanks!
This looks good, even though the performance impact does not seem huge.

andreasabel commented:
> `EmbPrj` does include a default implementation for `Enum`s already, which is used in a couple of places. There was some discussion about using them more widely in #5977 - I know there were concerns about forgetting to bump the interface version when changing enums.
> 
> It does feel a shame not to use the default methods here, but also I don't really have any good ideas about enforcing bumping the interface version - mostly noting this for context!

Forgetting to bump the interface version could lead to bad behaviours in case the old interface file is not malformed wrt. the new format.  Not sure how likely this is, but one scenario is that some developer inserts a new constructor for a `Enum` type in the middle rather than at the end.
Such bugs would affect users of `master` only (because in the release procedure we can include a mandatory bump of the interface version).

There was the suggestion to use the git commit hash as interface version.  This would lead to more frequent interface format changes and would negatively affect the performance for power users who are Agda developers or frequently updated Agda from `master`.

Atm, bumping the interface version is a manual process which can be forgotten even if changes to the `Serialize` modules where necessary.

Ideas how to improve the situation:
1. A PR template with a box to check: "In case you changed data types that are serialized (i.e. instantiate class `EmbPrj`), have you bumped the interface version at <point to source location>?"
2. Some sort of meta-program that queries for all data types that implement `EmbPrj` and hashes their sum-of-products representation.

SquidDev committed:
Encode enums as "raw" ints, rather than a node

This avoids one layer of indirection in serialisation/deserialisation,
slightly reducing the allocation rate and CPU time.

SquidDev committed:
Use valueN for single-constructor types

This has the same behaviour as vcase, but forms a more natural dual to
icodeN'.

andreasabel commented:
Thanks for splitting commits!
